src/amo/withExperiment.js

```diff
@@ -141,63 +154,80 @@ export const withExperiment = ({
+      _isExperimentEnabled: isExperimentEnabled,
       id: defaultId,
       variants: defaultVariants,
     };
 
     static displayName = `WithExperiment(${getDisplayName(WrappedComponent)})`;
 
-    constructor(props: withExperimentInternalProps) {
-      super(props);
+    getExperiments() {
+      const { cookies } = this.props;
+
+      return cookies.get(EXPERIMENT_COOKIE_NAME) || {};
+    }
+
+    cookieIncludesExperiment(registeredExperiments: RegisteredExpermients) {
+      const { id } = this.props;
+
+      return Object.keys(registeredExperiments).includes(id);
+    }
 
-      if (!this.isExperimentEnabled()) {
-        log.debug(`Experiment "${props.id}" is not enabled by config.`);
-        return;
+    render() {
+      const {
+        _config,
+        _getVariant,
+        _isExperimentEnabled,
+        cookies,
+        id,
+        variants,
+        ...props
+      } = this.props;
+
+      const isEnabled = _isExperimentEnabled({ _config, id });
+      const registeredExperiments = this.getExperiments();
+      const experimentInCookie = this.cookieIncludesExperiment(
+        registeredExperiments,
+      );
+      const experimentsToStore = { ...registeredExperiments };
+
+      // Clear any disabled experiments from the cookie.
+      let cleanupNeeded = false;
+      for (const experimentId of Object.keys(registeredExperiments)) {
+        if (!_isExperimentEnabled({ _config, id: experimentId })) {
+          delete experimentsToStore[experimentId];
+          cleanupNeeded = true;
+        }
       }
 
-      const { _getVariant, cookies, id, variants } = this.props;
+      // Do we need to record this experiment in the cookie?
+      const addExperimentToCookie = isEnabled && !experimentInCookie;
 
-      this.experimentCookie = cookies.get(this.getCookieName());
+      if (addExperimentToCookie) {
+        const variantToStore = _getVariant({ variants });
+        experimentsToStore[id] = variantToStore.id;
 
-      if (this.experimentCookie === undefined) {
-        const variant = _getVariant({ variants });
-        this.experimentCookie = variant.id;
-        cookies.set(this.getCookieName(), this.experimentCookie, cookieConfig);
-        if (variant.id !== NOT_IN_EXPERIMENT) {
+        if (variantToStore.id !== NOT_IN_EXPERIMENT) {
           // send an enrollment event
           _tracking.sendEvent({
-            action: variant.id,
+            action: variantToStore.id,
             category: [EXPERIMENT_ENROLLMENT_CATEGORY, id].join(' '),
           });
         }
       }
-    }
-
-    getCookieName() {
-      return `${this.props.id}Experiment`;
-    }
 
-    isExperimentEnabled() {
-      const { _config, id } = this.props;
-
-      return _config.get('experiments')[id] === true;
-    }
-
-    render() {
-      const { cookies, id, ...props } = this.props;
+      if (cleanupNeeded || addExperimentToCookie) {
+        cookies.set(EXPERIMENT_COOKIE_NAME, experimentsToStore, cookieConfig);
```

Comment:
it might be a good idea to do some sort of `sort()` on the value we store in the cookie, to avoid cases where the ids are the same but read in a different order for some reasons.

Reply:
The `cookies` library takes care of stringifying and parsing the data to and from JSON for us, so all we pass into it is an object. Properties of an object don't exist in a particular order, and cannot be sorted. JSON.stringify shouldn't result in different outputs of the same object at different times (although it's possible with different browsers), so I think what we have here is good enough.

If you think we need to do some sorting, how would you implement that? I guess we can covert the object to an array, sort it, store it, and then convert it from an array back to an object after we read it, but that seems like overkill to me.

Reply:
I think order of keys in a JS dict are preserved, but it's not a blocker and that was just an idea.